src/engine/AbsoluteValueConstraint.cpp

```diff
@@ -0,0 +1,454 @@
+                                 GlobalConfiguration::ABS_CONSTRAINT_COMPARISON_TOLERANCE );
+}
+
+List<PiecewiseLinearConstraint::Fix> AbsoluteValueConstraint::getPossibleFixes() const
+{
+    ASSERT( !satisfied() );
+    ASSERT( _assignment.exists( _b ) );
+    ASSERT( _assignment.exists( _f ) );
+
+    double bValue = _assignment.get( _b );
+    double fValue = _assignment.get( _f );
+
+    ASSERT( !FloatUtils::isNegative( fValue ) );
+
+    List<PiecewiseLinearConstraint::Fix> fixes;
+
+    // Possible violations:
+    //   1. f is positive, b is positive, b and f are unequal
+    //   2. f is positive, b is negative, -b and f are unequal
+
+    fixes.append( PiecewiseLinearConstraint::Fix( _b, fValue ) );
+    fixes.append( PiecewiseLinearConstraint::Fix( _b, -fValue ) );
+    fixes.append( PiecewiseLinearConstraint::Fix( _f, FloatUtils::abs( bValue ) ) );
+
+    return fixes;
+}
+
+List<PiecewiseLinearConstraint::Fix> AbsoluteValueConstraint::getSmartFixes( ITableau */* tableau */ ) const
+{
+    return getPossibleFixes();
+}
+
+List<PiecewiseLinearCaseSplit> AbsoluteValueConstraint::getCaseSplits() const
+{
+    if ( _phaseStatus != PhaseStatus::PHASE_NOT_FIXED )
+        throw MarabouError( MarabouError::REQUESTED_CASE_SPLITS_FROM_FIXED_CONSTRAINT );
+
+    List<PiecewiseLinearCaseSplit> splits;
+    splits.append( getNegativeSplit() );
+    splits.append( getPositiveSplit() );
+
+    return splits;
+}
+
+PiecewiseLinearCaseSplit AbsoluteValueConstraint::getNegativeSplit() const
+{
+    PiecewiseLinearCaseSplit negativePhase;
+
+    // Negative phase: b <= 0, b + f = 0
+    negativePhase.storeBoundTightening( Tightening( _b, 0.0, Tightening::UB ) );
+
+    Equation negativeEquation( Equation::EQ );
+    negativeEquation.addAddend( 1, _b );
+    negativeEquation.addAddend( 1, _f );
+    negativeEquation.setScalar( 0 );
+    negativePhase.addEquation( negativeEquation );
+
+    return negativePhase;
+}
+
+PiecewiseLinearCaseSplit AbsoluteValueConstraint::getPositiveSplit() const
+{
+    PiecewiseLinearCaseSplit positivePhase;
+
+    // Positive phase: b >= 0, b - f = 0
+    positivePhase.storeBoundTightening( Tightening( _b, 0.0, Tightening::LB ) );
+
+    //b - f = 0
+    Equation positiveEquation( Equation::EQ );
+    positiveEquation.addAddend( 1, _b );
+    positiveEquation.addAddend( -1, _f );
+    positiveEquation.setScalar( 0 );
+    positivePhase.addEquation( positiveEquation );
+
+    return positivePhase;
+}
+
+bool AbsoluteValueConstraint::phaseFixed() const
+{
+    return _phaseStatus != PhaseStatus::PHASE_NOT_FIXED;
+}
+
+PiecewiseLinearCaseSplit AbsoluteValueConstraint::getValidCaseSplit() const
+{
+    ASSERT( _phaseStatus != PhaseStatus::PHASE_NOT_FIXED );
+
+    if ( _phaseStatus == PhaseStatus::PHASE_POSITIVE )
+        return getPositiveSplit();
+
+    return getNegativeSplit();
+}
+
+void AbsoluteValueConstraint::eliminateVariable( unsigned /* variable */, double /* fixedValue */ )
```

Comment:
A general question, how can we eliminate variables?

Reply:
The preprocessor eliminates variables if they become fixed (5 <= x <= 5), or if identical varaibles are discovered (x=y; one is eliminated from all other equations).